doc/awl.1
.\" Generated by scdoc 1.11.2
.\" Complete documentation for this program is not available as a GNU info page
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.\" Begin generated content:
.TH "awl" "1" "2022-08-03"
.PP
.SH NAME
awl - DNS lookup tool
.PP
.SH SYNOPSIS
\fBawl\fR [ \fIOPTIONS\fR ] \fIname\fR [ \fI@server\fR ] [ \fItype\fR ]
.br
where
.PP
\fIname\fR is the query to make (\fBexample: froth.\&zone\fR)
.br
\fI@server\fR is the server to query (\fBexample: dns.\&froth.\&zone\fR)
.br
\fItype\fR is the DNS resource type (\fBexample: AAAA\fR)
.PP
.SH DESCRIPTION
.PP
\fBawl\fR  (\fBa\fRwls \fBw\fRant \fBl\fRicorice) is a simple tool designed to make DNS queries,
much like the venerable \fIdig\fR(1).\& An awl is a tool used to make small holes,
typically used in leatherworking.\&
.PP
\fBawl\fR is designed to be a more "modern" version of \fIdrill\fR(1)  by including 
some more recent RFCs and output options.\& \fBawl\fR is still heavily 
Work-In-Progress so some features may get added or removed.\&
.PP
.SH OPTIONS
.RS 4
Dig-like +[no]flags are supported, see dig(1)
.PP
\fB-D\fR, \fB--dnssec\fR, \fB+dnssec\fR
.br
	Enable DNSSEC.\& This needs to be manually enabled.\&
.PP
\fB-v\fR \fIvalue\fR
.br
	Set verbosity (currently WIP)
.PP
\fB-V\fR
.br
	Print the version and exit.\&
.PP
\fB-h\fR
.br
	Show a "short" help message.\&
.PP
.RE
.SS Query Options
.RS 4
\fB-4\fR
.br
	Only make query over IPv4
.PP
\fB-6\fR
.br
	Only make query over IPv6
.PP
\fB-p\fR, \fB--port\fR \fIport\fR
.br
	Sets the port to query.\&
.br
	
.br
\fIDefault Ports\fR:
.RS 4
.PD 0
.IP \(bu 4
\fI53\fR for \fBUDP\fR and \fBTCP\fR
.IP \(bu 4
\fI853\fR for \fBTLS\fR and \fBQUIC\fR
.IP \(bu 4
\fI443\fR for \fBHTTPS\fR
.PD
.PP
.RE
\fB-q\fR, \fB--query\fR \fIdomain\fR
.br
	Domain to query (eg.\& example.\&com)
.PP
\fB-c\fR, \fB--class\fR \fIclass\fR
.br
	DNS class to query (eg.\& IN, CH)
.PP
\fB-t\fR, \fB--qType\fR \fItype\fR
.br
	DNS type to query (eg.\& A, NS)
.PP
\fB--no-truncate\fR, \fB+ignore\fR
.br
	Ignore UDP truncation (by default, awl \fIretries with TCP\fR)
.PP
\fB--tcp\fR, \fB+tcp\fR, \fB+vc\fR
.br
	Use TCP for the query (see \fIRFC 7766\fR)
.PP
\fB--dnscrypt\fR, \fB+dnscrypt\fR
.br
	Use DNSCrypt
.PP
\fB-T\fR, \fB--tls\fR, \fB+tls\fR
.br
	Use DNS-over-TLS, implies \fB--tcp\fR (see \fIRFC 7858\fR)
.PP
\fB-H\fR.\& \fB--https\fR, \fB+https\fR
.br
	Use DNS-over-HTTPS (see \fIRFC 8484\fR)
.PP
\fB-Q\fR.\& \fB--quic\fR, \fB+quic\fR
.br
	Use DNS-over-QUIC (see \fIRFC 9250\fR)
.PP
\fB-x\fR, \fB--reverse\fR
.br
	Do a reverse lookup.\& Sets default \fItype\fR to PTR.\&
.br
	\fBawl\fR automatically makes an IP or phone number canonical.\&
.PP
\fB--timeout\fR \fIseconds\fR, \fB+timeout=\fR\fIseconds\fR
.br
	Set the timeout period.\& Floating point numbers are accepted.\&
.br
	0.\&5 seconds is the minimum.\&
.PP
\fB--retries\fR \fIint\fR, \fB+tries\fR=\fIint\fR, \fB+ retry\fR=\fIint\fR
.br
	Set the number of retries.\&
.br
	Retry is one more than tries, dig style
.PP
.RE
.SS DNS Flags
.PP
.RS 4
\fB--aa=[false]\fR, \fB+[no]aaflag\fR
.br
	(Set, Unset) AA (Authoritative Answer) flag
.PP
\fB--ad=[false]\fR, \fB+[no]adflag\fR
.br
	(Set, Unset) AD (Authenticated Data) flag
.PP
\fB--tc=[false]\fR, \fB+[no]tcflag\fR
.br
	(Set, Unset) TC (TrunCated) flag
.PP
\fB-z=[false]\fR, \fB+[no]zflag\fR
.br
	(Set, Unset) Z (Zero) flag
.PP
\fB--cd=[false]\fR, \fB+[no]cdflag\fR
.br
	(Set, Unset) CD (Checking Disabled) flag
.PP
\fB--qr=[false]\fR, \fB+[no]qrflag\fR
.br
	(Set, Unset) QR (QueRy) flag
.PP
\fB--rd=[true]\fR, \fB+[no]rdflag\fR
.br
	(Set, Unset) RD (Recursion Desired) flag
.PP
\fB--ra=[false]\fR, \fB+[no]raflag\fR
.br
	(Set, Unset) RA (Recursion Available) flag
.PP
.RE
.SS Output Display
.RS 4
\fB--no-question\fR, \fB+noquestion\fR
.br
	Do not display the Question section
.PP
\fB--no-answer\fR, \fB+noanswer\fR
.br
	Do not display the Answer section
.PP
\fB--no-answer\fR, \fB+noanswer\fR
.br
	Do not display the Answer section
.PP
\fB--no-authority\fR, \fB+noauthority\fR
.br
	Do not display the Authority section
.PP
\fB--no-additional\fR, \fB+noadditional\fR
.br
	Do not display the Additional section
.PP
\fB--no-statistics\fR, \fB+nostats\fR
.br
	Do not display the Statistics (additional comments) section
.PP
.RE
.SS Output Formats
.RS 4
\fB-j\fR, \fB--json\fR, \fB+json\fR
.br
	Print the query results as JSON.\&
.PP
\fB-X\fR, \fB--xml\fR, \fB+xml\fR
.br
	Print the query results as XML.\&
.PP
\fB-y\fR, \fB--yaml\fR, \fB+yaml\fR
.br
	Print the query results as YAML.\&
.PP
\fB-s\fR, \fB--short\fR, \fB+short\fR
.br
	Print just the address of the answer.\&
.PP
.RE
.SH EXAMPLES
.nf
.RS 4
awl grumbulon\&.xyz -j +cd
.fi
.RE
Run a query of your local resolver for the A records of grumbulon.\&xyz, print
them as JSON and disable DNSSEC verification.\&
.PP
.nf
.RS 4
awl +short example\&.com AAAA @1\&.1\&.1\&.1
.fi
.RE
Query 1.\&1.\&1.\&1 for the AAAA records of example.\&com, print just the answers
.PP
.nf
.RS 4
awl -xT PTR 8\&.8\&.4\&.4 @dns\&.google
.fi
.RE
Query dns.\&google over TLS for the PTR record to the IP address 8.\&8.\&4.\&4
.PP
.SH SEE ALSO
\fIdrill\fR(1), \fIdig\fR(1), the many DNS RFCs

awl.ts
// SPDX-License-Identifier: MIT
import { bold, italic, parse, underline } from "./deps.ts";
import { QueryResponse, ServerOptions } from "./lib/utils.ts";
import { doQuery } from "./lib/query.ts";
import { parseResponse } from "./lib/response.ts";
import { parseArgs } from "./args.ts";
import { parseNAPTR, parsePTR } from "./lib/reverse.ts";

async function main() {
  // Parse args
  const args = parse(Deno.args, {
    alias: {
      h: "help",
      p: "port",
      s: "short",
      x: "ptr",
      V: "version",
    },
    boolean: ["help", "ptr", "short", "version"],
    string: ["port"],
    default: {
      "port": "53",
    },
  });
  if (args.help) args.version = true;

  if (args.version) {
    console.log(
      `${
        bold("awl")
      } version 0.1.0 (running with deno ${Deno.version.deno}, TypeScript ${Deno.version.typescript}, on V8 ${Deno.version.v8})
Written by (YOUR NAME GOES HERE)`,
    );
  }

  if (args.help) {
    console.log(
      ` ${bold("Usage:")} awl name ${italic("type")} ${italic("@server")}
       ${bold("<name>")}    domain name
       ${bold("<type>")}    defaults to A
       ${bold("<@server>")} defaults to your local resolver

       Order ${bold("DOES NOT")} matter\n
      `,
      `${underline("Options")}:
        -p <port> use <port> for query, defaults to 53
        -s Equivalent to dig +short, only return addresses
        -x do a reverse lookup
        -h print this helpful guide
        -V get the version
      `,
    );
  }

  if (args.version) Deno.exit(0);

  const parsedArgs = parseArgs(args);

  let domain = parsedArgs.name || ".";

  let query = parsedArgs.type || "A";
  if (domain === ".") query = "NS";

  if (query === "PTR") {
    // The "server" is an IP address, it needs to become a canonical domain
    domain = parsePTR(domain);
  } else if (query === "NAPTR") {
    // The "server" is a phone number, it needs to become a canonical domain
    domain = parseNAPTR(domain);
  }

  const server = parsedArgs.server || { server: "", port: 53 };

  const response: QueryResponse = await doQuery(domain, query, server);

  if (!args.short) {
    console.log(
      `;; ->>HEADER<<- opcode: QUERY, rcode: ${response.response}
;; QUESTION SECTION:
;; ${domain}    IN    ${query}
  
;; ANSWER SECTION:`,
    );
  }
  if (response.response === "NOERROR") {
    const res = parseResponse(response, domain, query, args.short);
    res.forEach((answer) => {
      console.log(answer);
    });
  }
  if (!args.short) {
    console.log(`
;; ADDITONAL SECTION:

;; Query time: ${response.time} msec
;; SERVER: ${displayServer(server)}
;; WHEN: ${new Date().toLocaleString()}
        `);
  }
}

/**
 * A handler for displaying the server
 * @param {ServerOptions} server - The DNS server used
 * @returns {string} The string used
 */
function displayServer(server: ServerOptions): string {
  let val = "";
  if (server.server) {
    val += server.server;
  }
  if (server.port != 53) {
    val += `#${server.port}`;
  }
  if (!val) val = "System";
  return val;
}

if (import.meta.main) {
  await main();
}
